Model/ResourceModel/NodeType/Category.php

```diff
@@ -101,4 +102,22 @@ public function fetchData($storeId = Store::DEFAULT_STORE_ID, $categoryIds = [])
 
         return $connection->fetchPairs($select);
     }
+
+    /**
+     * Get products count in categories
+     *
+     * @see \Magento\Catalog\Model\ResourceModel\Category::getProductCount
+     */
+    public function getCategoriesProductCount($categoryIds = [])
+    {
+        $productTable = $this->getConnection()->getTableName('catalog_category_product');
```

Comment:
This may provide misleading data as `catalog_category_product` table is source of truth for all product category associations including disabled and not visible products. (In Example will indicate category full of disabled products as non-zero products, even thou navigating to that category will yield zero products).

There is table `catalog_category_product_index_store<store_id>` that contains visibility - can You check if it is depopluated when changing product status? As it may be better source of data if it is indeed not having entries for disabled products.

Reply:
Just checked with all indexers set to "On Save" mode.

All entries for e.g. product id 2040 disappeared from `catalog_category_product_index_store1` after disabling the product. Re-enabling product re-populated the entries.